pkg/startup/startup.go

```diff
@@ -687,11 +715,35 @@ func evaluateENVBool(envVar string, defaultValue bool) bool {
 
 // autoDetectCIDR auto-detects the IP and Network using the requested
 // detection method.
-func autoDetectCIDR(method string, version int) *cnet.IPNet {
+func autoDetectCIDR(method string, version int,
+	k8sClient kubernetes.Interface, nodeName string) *cnet.IPNet {
+	log.Infof("IP_AUTODETECTION_METHOD=%s", method)
+
 	if method == "" || method == AUTODETECTION_METHOD_FIRST {
 		// Autodetect the IP by enumerating all interfaces (excluding
 		// known internal interfaces).
 		return autoDetectCIDRFirstFound(version)
+	} else if method == AUTODETECTION_METHOD_K8S {
+		if k8sClient == nil {
+			log.Errorf("Invalid k8s client for IP autodetection method: %s", method)
+			return nil
+		}
+
+		k8sNode, err := k8sClient.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("Failed to get the k8s node for name %q", nodeName)
+			return nil
+		}
+
+		ips := autodetection.K8sNodeInternalIPs(k8sNode, version)
+		if len(ips) == 0 {
+			log.Errorf("No k8s internal IP version %d", version)
+			return nil
+		}
+		if len(ips) > 1 {
+			log.Warnf("More then 1 k8s internal IP version %d, picking one: %s", version, ips[0])
```

Comment:
Do we have any future plans to improve on this? Do we need / want a way to limit the address selection even further?

Reply:
The future plan we pondered with @fasaxc is to introduce a field (in the next step to add external IP) that could hold any number of addresses and keep the current single-address field for compatibility. That would also address the issue with possible multiple internal IPs in k8s.

This PR would make IP detection in k8s deterministic with minimal change and the rest being a gradual improvement on it.

Reply:
That is probably a good thing to do. It is an API change though so we need to be careful about spelling out exactly how the field looks and behaves in various scenarios, how upgrade works, etc.  

Personally, I'd prefer if we have that design agreed before we make intermediate steps towards it so that we don't make any missteps.